Adding caching, case handling, more

navijation opened the pull request:
1: Added caching of certain variables that are computationally heavy to compute and might be computed in quick succession in a graphing algorithm. This has to be tested, but it's fairly simple. See the comments on the cached variables and refreshCache() for implementation details.
2: implemented Iterable<PeerInteraction> interface
3: Added a pushBack() method that allows PeerInteractions to be inserted in unsorted order (just to weaken preconditions and the insidious bugs that can arise from breaking them)
4: Added a mergeAllDuplicates() and a mergeRecentDuplications() method [I said I'd do this a while ago but never got to it]. I'm not sure which to keep right now. The former is much safer but can be unnecessarily slower depending on usage patterns. The reason I don't run these in the addEntry() method is that merging entries sequentially will have unwanted formatting effects (see PeerInteraction(PeerInteraction[]) for details).They are meant to be merged as batches.
5: Added a test main method. This currently tests merging, inserting unsorted data, and throwing exceptions for ID mismatches.
6: Fixed a silly bug in PeerInteraction merge constructor (didn't copy dates properly). Also modified behavior slightly to be more forgiving.

navijation committed:
Adding caching, case handling, more

1: Added caching of certain variables that are computationally heavy to compute and might be computed in quick succession in a graphing algorithm. This has to be tested, but it's fairly simple. See the comments on the cached variables and refreshCache() for implementation details.
2: implemented Iterable<PeerInteraction> interface
3: Added a pushBack() method that allows PeerInteractions to be inserted in unsorted order (just to weaken preconditions and the insidious bugs that can arise from breaking them)
4: Added a mergeAllDuplicates() and a mergeRecentDuplications() method [I said I'd do this a while ago but never got to it]. I'm not sure which to keep right now. The former is much safer but can be unnecessarily slower depending on usage patterns. The reason I don't run these in the addEntry() method is that merging entries sequentially will have unwanted formatting effects (see PeerInteraction(PeerInteraction[]) for details).They are meant to be merged as batches.
5: Added a test main method. This currently tests merging, inserting unsorted data, and throwing exceptions for ID mismatches.

navijation committed:
Fixed merge constructor bug

Forgot to merge dates. Also, now the merge constructor will use the last valid entry instead of the last entry as a reference for feedback information (in case someone pressed submit twice or something on accident, somehow).

NavneethJayendran commented:
I'm not sure where the compilation error is. I most likely am using an outdated method somewhere. Is Lecture.get(PeerInteraction) still there?

navijation committed:
Large changes to Student class

Large changes.to Student, summarized as:
1) Added caching for computationally nontrivial methods that may get called repeatedly in a short time.
2) Implemented Iterable<PeerInteraction>
4) Added mergeAllDuplicates() and mergeRecentDuplicates() [latter seems useful after adding a single batch of entries from a single lecture,
    former seems more useful for testing and is generally safer but slower] 
5) Main method used to test performance of mergeAllDuplicates().

navijation committed:
Synchronized Roster with super-branch

I accidentally committed these changes to the super-branch instead of to this one... so I'm basically copying these changes over.

navijation committed:
Added remove() to StudentIterator

On my machine this compiled without a problem, but Travis CI doesn't agree with omission of this method (are they using an older version of Java than 1.8, perhaps?). This should now be corrected.

navijation committed:
Changed LectureInitializer ref to LectureData

Accidentally pasted the unfixed version of this file at first. Now all should be fixed.

NavneethJayendran commented:
At this point I'm fairly certain that this should compile... I feel like there's something wrong with the Travis CI's settings. I don't really know how it works, so I would appreciate it if @Togira could check this out whenever he's free. I'll get to resolving merge conflicts soon. For now, inspect the changes if you like.

Commit:
Fixed compiler errors by reverting the roster class back to what I think it should be.

Commit:
Merge branch 'update-PeerInteraction-test' into caching-merging-etc
